### PR TITLE
gather elasticsearch indices and shard stats

### PR DESCRIPTION
**Pull request by mhohara**
### Required for all PRs:

- [x ] Signed [CLA](https://influxdata.com/community/cla/).
- [x ] Associated README.md updated.
- [x ] Has appropriate unit tests.

### CODE REVIEW

**Comment by mhohara** [comment]
@danielnelson  Much easier with a new fork. I'd accumulated too much while the old one was open over the last 8 months. Thanks for your patience.

**Comment by mhohara** [comment]
@danielnelson  The circle test failed, but unless I'm reading it incorrectly, its unrelated to any changes of mine.  Specifically, the inputs/elasticsearch test did complete OK. The only error I noted was:

--- FAIL: TestMemcachedGeneratesMetrics (0.00s)
	Error Trace:	memcached_test.go:25
	Error:      	Received unexpected error:
	            	dial tcp 127.0.0.1:11211: getsockopt: connection refused
FAIL
FAIL	github.com/influxdata/telegraf/plugins/inputs/memcached	0.022s

File: plugins/inputs/elasticsearch/elasticsearch.go [hunk 1]
```diff
@@ -274,7 +299,10 @@ func (e *Elasticsearch) gatherClusterHealth(url string, acc telegraf.Accumulator
 		return err
 	}
 	measurementTime := time.Now()
+	tags := map[string]string{"name": healthStats.ClusterName}
+
 	clusterFields := map[string]interface{}{
+		"cluster_name":          healthStats.ClusterName,
```
Comment (lpic10) [inline plugins/inputs/elasticsearch/elasticsearch.go]: If I understand correctly this is adding a new field named `cluster_name` after having changed the previous tag `cluster_name` to `name` with the exact same information. Is there any reason for this?

File: plugins/inputs/elasticsearch/elasticsearch.go [hunk 2]
```diff
@@ -288,12 +316,14 @@ func (e *Elasticsearch) gatherClusterHealth(url string, acc telegraf.Accumulator
...
 		clusterFields,
-		map[string]string{"name": healthStats.ClusterName},
+		tags,
 		measurementTime,
 	)
 
 	for name, health := range healthStats.Indices {
+		tags := map[string]string{"name": healthStats.ClusterName, "index_name": name}
 		indexFields := map[string]interface{}{
+			"index_name":            name,
```
Comment (lpic10) [inline plugins/inputs/elasticsearch/elasticsearch.go]: any reason to have `index_name` as field and as a tag at same time?

File: plugins/inputs/elasticsearch/elasticsearch.go (hunk 1 above)
Comment (mhohara) [inline plugins/inputs/elasticsearch/elasticsearch.go]: Presumably better to simply contain this information in the cluster_name tag.

File: plugins/inputs/elasticsearch/elasticsearch.go (hunk 2 above)
Comment (mhohara) [inline plugins/inputs/elasticsearch/elasticsearch.go]: I can get rid of the redundant (non-tag) index_name.

**Comment by lpic10** [comment]
@mhohara I will try to test this soon, you need to update the README after the latest changes

**Comment by mhohara** [comment]
@lpic10  Readme updated. And thanks for looking at this.

**Comment by eran-totango** [comment]
@mhohara @lpic10 any ETA for the indicestats feature?

**Comment by mhohara** [comment]
@lpic10 Any progress on your review?

**Comment by florck** [comment]
👍

**Comment by mhohara** [comment]
@danielnelson @lpic10  Any progress here? I don't see much in the way of other updates to the ES plugin as of yet.

**Comment by danielnelson** [comment]
Sorry no progress to report, we still need to do #2650 #2711 first.

**Comment by mhohara** [comment]
Are you waiting for something from me on those two?

**Comment by danielnelson** [comment]
No, but you are certainly welcome to give fixing #2650 a shot.

**Comment by mhohara** [comment]
@danielnelson I'm not sure I understand the issue in #2650 . Could the e.masterNodeId from the catmaster be compared to a cat/nodeattrs call to localhost:9200 to get the NodeId of the node its on,  to see whether 'it' is the master?

**Comment by danielnelson** [comment]
The problem is that there is a race condition, each server is collected concurrently via the anonymous goroutine in Gather, so the order in which e.isMaster is assigned too [here](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/elasticsearch/elasticsearch.go#L241) is not defined.  The isMaster value should probably not be a field on the plugin struct.

**Comment by mhohara** [comment]
In this PR's version of the code, the isMaster flag is only used in the Gather routine. Perhaps the necessary ES API calls to determine whether the node that telegraf is running on is the cluster master need to be put into the Gather routine?

**Comment by danielnelson** [comment]
I remember the PR code did not fix the problem, the routine that determines master status is less important than not storing it on the struct where it will be available/overwritten by other routines.

**Comment by mhohara** [comment]
The flag was put in the struct because it is the only output of the gatherNodeStats routine which did the ES API calls to get the node information and hence could decide whether a node was master. So it seemed the best/only way to return the isMaster information back to the calling Gather routine. Moving 
 the call needed to decide whether the node is master to the Gather routine seems a reasonable way to fix the issue (if I understand the issue correctly).

**Comment by hgfischer** [comment]
Is this abandoned? Do you need help? I am interested in this feature.

**Comment by danielnelson** [comment]
@hgfischer We need to address #2650 and #2711 before completing this issue.

**Comment by danielnelson** [comment]
Closed in #6060

**Pull request closed without merging**